lib/galaxy/tools/imp_exp/unpack_tar_gz_archive.py

```diff
@@ -35,21 +36,33 @@ def check_archive(archive_file, dest_dir):
     Ensure that a tar archive has no absolute paths or relative paths outside
     the archive.
     """
-    with tarfile.open(archive_file, mode="r") as archive_fp:
-        for arc_path in archive_fp.getnames():
-            assert os.path.normpath(os.path.join(dest_dir, arc_path)).startswith(
-                dest_dir.rstrip(os.sep) + os.sep
-            ), f"Archive member would extract outside target directory: {arc_path}"
+    if zipfile.is_zipfile(archive_file):
+        with zipfile.ZipFile(archive_file, "r") as archive_fp:
+            for arc_path in archive_fp.namelist():
+                assert not os.path.isabs(arc_path), f"Archive member has absolute path: {arc_path}"
+                assert not os.path.relpath(arc_path).startswith(
+                    ".."
+                ), f"Archive member would extract outside target directory: {arc_path}"
+    else:
+        with tarfile.open(archive_file, mode="r") as archive_fp:
+            for arc_path in archive_fp.getnames():
+                assert os.path.normpath(os.path.join(dest_dir, arc_path)).startswith(
+                    dest_dir.rstrip(os.sep) + os.sep
+                ), f"Archive member would extract outside target directory: {arc_path}"
     return True
 
 
 def unpack_archive(archive_file, dest_dir):
     """
     Unpack a tar and/or gzipped archive into a destination directory.
     """
-    archive_fp = tarfile.open(archive_file, mode="r")
-    archive_fp.extractall(path=dest_dir)
-    archive_fp.close()
+    if zipfile.is_zipfile(archive_file):
```

Comment:
Wondering of functionality from [compression_utils.py](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/util/compression_utils.py) might be useful here?

Reply:
Thanks, I can certainly try to refactor it to use compression utils, I just hope refactoring it won't break the old import :crossed_fingers:

Reply:
maybe just one thing at a time :sweat_smile:

Reply:
Ok, I've created #15759 for tracking the refactoring against dev instead of trying to squeeze it in the bug fix. It certainly feels safer.